src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,284 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
     vt->return_padata = pkinit_server_return_padata;
     return 0;
 }
+
+/*
+ * Run the received, verified certificate through the certauth plugin interface,
+ * to verify that client can use the certificate based on the extension policy
+ * defined by modules.
+ */
+static krb5_error_code
+server_authorize_cert(krb5_context context, certauth_handle *certauth_modules,
+                      pkinit_kdc_context plgctx,
+                      pkinit_kdc_req_context reqctx,
+                      krb5_kdcpreauth_callbacks cb, krb5_kdcpreauth_rock rock,
+                      krb5_principal client)
+{
+    krb5_error_code ret;
+    certauth_handle hd;
+    struct certauth_req_opts opts;
+    krb5_boolean accepted = FALSE;
+    krb5_data cert;
+    size_t i;
+    void *db_ent = NULL;
+    char **ais = NULL, **ai = NULL;
+
+    /* Re-encode the received certificate into DER, which is extra work, but
+     * avoids creating a crypto dependency on the interface. */
+    ret = crypto_encode_der_cert(context, reqctx->cryptoctx, &cert);
+    if (ret)
+        goto cleanup;
+
+    /* Set options for the builtin module. */
+    opts.plgctx = plgctx;
+    opts.reqctx = reqctx;
+    opts.cb = cb;
+    opts.rock = rock;
+
+    db_ent = cb->client_entry(context, rock);
+
+    /*
+     * Check the certificate against each certauth module. For the certificate
+     * to be authorized at least one module must return 0 and no module returns
+     * an error code other than KRB5_PLUGIN_NO_HANDLE (pass).  Add indicators
+     * from modules that return 0 or pass.
+     */
+    ret = KRB5_PLUGIN_NO_HANDLE;
+    for (i = 0; certauth_modules != NULL && certauth_modules[i] != NULL; i++) {
+        hd = certauth_modules[i];
+        if (hd->vt.authorize == NULL)
+            continue;
+
+        ret = hd->vt.authorize(context, hd->moddata, cert, client, &opts,
+                               db_ent, &ais);
+
+        if (ret == 0)
+            accepted = TRUE;
+        else if (ret != KRB5_PLUGIN_NO_HANDLE)
+            goto cleanup;
+
+        if (ais != NULL) {
+            /* Assert authentication indicators from certificate authorization. */
+            for (ai = ais; *ai != NULL; ai++) {
+                ret = cb->add_auth_indicator(context, rock, *ai);
+                if (ret)
+                    goto cleanup;
+            }
+
+            if (hd->vt.free_ind != NULL)
+                hd->vt.free_ind(context, hd->moddata, ais);
+            else
+                free_ai_list(ais);
+            ais = NULL;
+        }
+    }
+
+    if (accepted)
+        ret = 0;
```

Comment:
And otherwise we return KRB5_PLUGIN_NO_HANDLE?  I don't think that's meaningful to the caller.  I think we should return KRB5KDC_ERR_CLIENT_NAME_MISMATCH, in that no module has asserted a match between the certificate and the client principal name.